// grpc-helper/dist/src/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var promClient = require("prom-client");
var grpc = require("grpc");
var histogram = new promClient.Histogram({
    name: 'grpc_response_duration_seconds',
    help: 'Histogram of grpc response in seconds',
    labelNames: ['peer', 'method', 'code'],
});
function getMetricsInterceptor() {
    return function metricsInterceptor(options, nextCall) {
        var call = nextCall(options);
        var endTimer = histogram.startTimer({
            peer: call.getPeer(),
            method: options.method_definition.path,
        });
        var requester = (new grpc.RequesterBuilder())
            .withStart(function (metadata, _listener, next) {
            var newListener = (new grpc.ListenerBuilder())
                .withOnReceiveStatus(function (status, next) {
                endTimer({
                    code: status.code,
                });
                next(status);
            }).build();
            next(metadata, newListener);
        }).build();
        return new grpc.InterceptingCall(call, requester);
    };
}
exports.getMetricsInterceptor = getMetricsInterceptor;
